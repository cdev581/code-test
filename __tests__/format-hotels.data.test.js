import hotelsData from "../src/data/hotels.json";
import formatHotelsData from "../src/lib/format-hotels-data";

beforeEach(() => {
  global.structuredClone = jest.fn((obj) => JSON.parse(JSON.stringify(obj)));
});

afterEach(() => {
  jest.restoreAllMocks();
});

describe("format hotels lib", () => {
  test("hotel item formatting produces object of type Hotel", () => {
    const hotels = formatHotelsData(hotelsData);
    expect(hotels).toMatchSnapshot();
  });

  test("formatting wrong result type as hotelData produces empty array", () => {
    const hotels = formatHotelsData({results: 10});
    expect(hotels).toEqual([]);
  });

  test("formatting a hotel with broken image link gives default image", () => {
    const cloned = global.structuredClone(hotelsData);
    cloned.results[0].property.previewImage.url = '!';
    const hotels = formatHotelsData(cloned);
    expect(hotels[0].image.src).toEqual("/blank.svg");
  });

  test("formatting a hotel with missing properties defaults gracefully", () => {
    const hotels = formatHotelsData({
      results: [
        {
          "property": {
            "previewImage": {
              "url": "https://unsplash.it/145/125/?random"
            },
            "rating": {
              "ratingType": "self"
            }
          },
          "offer": {
            "promotion": {
            },
            "displayPrice": {
              "currency": "AUD"
            },
            "savings": {
              "currency": "AUD"
            },
            "cancellationOption": {
              "cancellationType": "NOT_REFUNDABLE"
            }
          }
        }
      ]
    });
    expect(hotels).toMatchSnapshot();
  });
});