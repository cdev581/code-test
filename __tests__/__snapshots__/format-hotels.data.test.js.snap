// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format hotels lib formatting a hotel with missing properties defaults gracefully 1`] = `
[
  {
    "address": "",
    "cancellationType": "NOT_REFUNDABLE",
    "href": "https://qantas.com.au",
    "id": null,
    "image": {
      "alt": "",
      "src": "https://unsplash.it/145/125/?random=0",
    },
    "offerName": "",
    "price": null,
    "promotionTitle": null,
    "rating": null,
    "savings": null,
    "title": "",
  },
]
`;

exports[`format hotels lib hotel item formatting produces object of type Hotel 1`] = `
[
  {
    "address": "7-11 Talavera Rd, North Ryde",
    "cancellationType": "NOT_REFUNDABLE",
    "href": "https://qantas.com.au",
    "id": "cxd650nuyo",
    "image": {
      "alt": "Image of Courtyard by Marriott Sydney-North Ryde",
      "src": "https://unsplash.it/145/125/?random=0",
    },
    "offerName": "Deluxe Balcony Room",
    "price": {
      "amount": 329,
      "currency": "AUD",
    },
    "promotionTitle": "Exclusive Deal",
    "rating": {
      "ratingType": "self",
      "ratingValue": 4.5,
    },
    "savings": {
      "amount": 30,
      "currency": "AUD",
    },
    "title": "Courtyard by Marriott Sydney-North Ryde",
  },
  {
    "address": "339 Pitt St, Sydney",
    "cancellationType": "FREE_CANCELLATION",
    "href": "https://qantas.com.au",
    "id": "mesq6mggyn",
    "image": {
      "alt": "Image of Primus Hotel Sydney",
      "src": "https://unsplash.it/145/125/?random=1",
    },
    "offerName": "Deluxe King",
    "price": {
      "amount": 375,
      "currency": "AUD",
    },
    "promotionTitle": "Exclusive Deal",
    "rating": {
      "ratingType": "self",
      "ratingValue": 5,
    },
    "savings": {
      "amount": 28,
      "currency": "AUD",
    },
    "title": "Primus Hotel Sydney",
  },
  {
    "address": "389 Pitt St, Sydney",
    "cancellationType": "NOT_REFUNDABLE",
    "href": "https://qantas.com.au",
    "id": "xbtlihs45t",
    "image": {
      "alt": "Image of property",
      "src": "https://unsplash.it/145/125/?random=2",
    },
    "offerName": "Deluxe King Room",
    "price": {
      "amount": 227,
      "currency": "AUD",
    },
    "promotionTitle": "Red Hot",
    "rating": {
      "ratingType": "star",
      "ratingValue": 4,
    },
    "savings": null,
    "title": "Rydges World Square Sydney",
  },
  {
    "address": "150 Day Street, Sydney",
    "cancellationType": "FREE_CANCELLATION",
    "href": "https://qantas.com.au",
    "id": "5lm8loqk1s",
    "image": {
      "alt": "Image of PARKROYAL Darling Harbour Sydney",
      "src": "https://unsplash.it/145/125/?random=3",
    },
    "offerName": "Deluxe King",
    "price": {
      "amount": 535,
      "currency": "AUD",
    },
    "promotionTitle": "Exclusive Deal",
    "rating": {
      "ratingType": "star",
      "ratingValue": 4.5,
    },
    "savings": null,
    "title": "PARKROYAL Darling Harbour Sydney",
  },
  {
    "address": "431-439 Pitt Street, Sydney",
    "cancellationType": "FREE_CANCELLATION",
    "href": "https://qantas.com.au",
    "id": "kwjf8jlxg9",
    "image": {
      "alt": "Image of Metro Hotel Marlow Sydney Central",
      "src": "https://unsplash.it/145/125/?random=4",
    },
    "offerName": "Deluxe King",
    "price": {
      "amount": 295,
      "currency": "AUD",
    },
    "promotionTitle": "Bonus Points",
    "rating": {
      "ratingType": "star",
      "ratingValue": 3.5,
    },
    "savings": null,
    "title": "Metro Hotel Marlow Sydney Central",
  },
]
`;
